### PR TITLE
Implementation for Decimal type with memory layout compatible with pyarrow's memory layout.

### PR DESCRIPTION
**Pull request by vasil-pashov**
* Load decimal from string
* Convert decimal to string
* Check for equality
* Add unit tests

Implementation is for 128 bit fixed point decimal. The number is stored as two little endian uint64_t variables. The format does not store the position of the decimal point.

TODO:
* Add arithmetic
* Add comparison operators

### CODE REVIEW

File: cpp/arcticdb/util/decimal.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+                const Decimal negated = this->operator-();
+                std::string result = negated.to_string(scale);
+                result.insert(0, "-");
+                return result;
+            }
+            if(data_[0] == 0 && data_[1] == 0) {
+                return "0";
+            }
+
+            boost::multiprecision::uint128_t number = to_uint128(
```
Comment (vasil-pashov) [inline cpp/arcticdb/util/decimal.cpp:277]: This also uses boost's implementation as intermediate type. Again we could change this if needed at the cost of having a bit more complicated code.

File: cpp/arcticdb/util/preprocess.hpp [hunk 2]
```diff
@@ -7,11 +7,6 @@
 
 #pragma once
 
-#define stringify(x) #x
-#define static_to_string(x) stringify(x)
-#define source_loc __FILE__ ":" static_to_string(__LINE__)
-#define to_id(x) ([]() constexpr { return x; })
```
Comment (vasil-pashov) [inline cpp/arcticdb/util/preprocess.hpp:13]: This was not used anywhere but for some reason makes the build to fail.

File: cpp/arcticdb/util/decimal.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            return digits_.data();
+        }
+
+        int NumberComponents::get_size() const {
+            return size_;
+        }
+
+        Decimal::Decimal() : data_{ 0 } {}
+
+        Decimal::Decimal(std::string_view number) : data_{0} {
```
Comment (vasil-pashov) [inline cpp/arcticdb/util/decimal.cpp:235]: I went with custom linear parsing, but regex might be a better idea. Any opinions?

**Review by willdealtry — CHANGES_REQUESTED** [review]

**Review by willdealtry — APPROVED** [review]

File: cpp/arcticdb/util/decimal.cpp [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+            }
+            if(data_[0] == 0 && data_[1] == 0) {
+                return "0";
+            }
+
+            boost::multiprecision::uint128_t number = to_uint128(
+                    data_[MOST_SIGNIFICANT_WORD_INDEX],
+                    data_[LEAST_SIGNIFICANT_WORD_INDEX]);
+
+            std::string result;
```
[the remark this answers is not in the document]
Comment (vasil-pashov) [reply cpp/arcticdb/util/decimal.cpp:281]: Yes, it's possible. I was thinking that for most of the cases we have the SSO will kick in. `sizeof(std::string)` is 32bytes, so about 30 chars, should fit it without allocation. If we do `result.reserve` we will be allocating always (I think).

**Pull request merged**